### PR TITLE
fix(inputs.postgres*)!: Prevent leaking sensitive data in server tag

### PR DESCRIPTION
**Pull request by srebhan**
## Summary

The current implementation of `SanitizedAddress()` does not handle quoted parameter values containing spaces or equal signs (e.g. `host=localhost user=john password='a secret=good'`). Additionally to those issues, the parsing of URI connection strings is completely broken for spaces and equal signs as values containing those need to be single-quoted  which does not happen. Furthermore, that function escapes "special characters" (space, backslash and single-quote) which subsequently breaks the sanitize function.

This PR adds unit-tests for both of the mentioned functions. It also removes the useless escaping of spaces during URI handling and replaces this part by quoting the value. Finally, the PR improves the regular-expression used to remove sensitive data to correctly handle quoted values, equal-signs and other corner cases.

## Checklist

- [x] No AI generated code was used in this PR

### CODE REVIEW

**Comment by james-johnston-thumbtack** [comment]
I am the person who originally reported this issue.  I didn't look closely at the updated implementation, but glancing through the new test cases, it makes sense to me and seems to be working.

**Review by powersj — COMMENTED** [review]
I am of course weary when we make these larger changes to fix an issue. Can you please add a note the change log about this change as it is breaking tags for some small percentage of users.

I would also like to land this in 1.30 and not 1.29.5. Agreed?

**Review by powersj — APPROVED** [review]

**Review by DStrand1 — APPROVED** [review]

**Review by jdstrand — CHANGES_REQUESTED** [review]
I think the PR looks fine overall. I didn't desk check the sanitizer but the existing tests show it is working better. Marking as 'request changes' since I'd like to see even more tests for this (there is an opportunity to refactor to make that better as well as a question on zero-ing out `addr`).

File: plugins/inputs/postgresql/service.go [hunk 1]
```diff
@@ -148,30 +152,32 @@ func (p *Service) Stop() {
...
+		if addr, err = toKeyValue(addr); err != nil {
+			return "", err
 		}
-	} else {
-		canonicalizedAddress = addr.String()
 	}
 
-	return kvMatcher.ReplaceAllString(canonicalizedAddress, ""), nil
+	// Sanitize the string using a regular expression
+	sanitized := sanitizer.ReplaceAllString(addr, "")
```
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/postgresql/service.go:179]: The problem is that we use-up a lot (i.e. three) of locked pages when "getting" secrets, that's why we want to hold them as "short-as-possible". If you are really concerned about holding those, I would switch to "TemporaryString()" but this has the risk of Telegraf using up the locked pages and in turn running into a panic...

File: plugins/inputs/postgresql/postgresql_test.go [hunk 2]
```diff
@@ -316,3 +317,211 @@ func TestPostgresqlDatabaseBlacklistTestIntegration(t *testing.T) {
...
+			name:  "simple text",
+			value: `foo`,
+		},
+		{
+			name:  "empty values",
+			value: ``,
+		},
+		{
+			name:  "space in value",
+			value: `foo bar`,
```
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/postgresql/postgresql_test.go:496]: Can you please provide an example input for those tests? Please note: you examples include invalid PostgreSQL DSNs which never end up in this function, so I'm a bit hesitant to add those...

**Comment by srebhan** [comment]
@jdstrand I hopefully addressed your comments. If I got things wrong, please let me know!

File: plugins/inputs/postgresql/service.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply plugins/inputs/postgresql/service.go:179]: @srebhan - I see in the updated commits that you are using `TemporaryString()`. This is of course fine with me from a security standpoint, but I was prepared to acquiesce to not using it. I don't recall all the details from before about when we'd run out of locked pages, but I was hoping that in this case it would be a quick open and close and not a leak of locked pages. This is a database plugin and I know that they tend to hold on to things.... I'll let you decide how to handle it now that I've rambled long enough, heh. :)

**Review by jdstrand — APPROVED** [review]
Thanks for the updates! I have one separate comment about reverting 6f322d31e282ee086aaec2b5dbbb843a34586026 if you think it will cause a problem. Approving as is and feel free to commit if you revert 6f322d31e282ee086aaec2b5dbbb843a34586026.

**Comment by srebhan** [comment]
@jdstrand I don't think using the temporary string is a problem, so let's try as-is.
